students/Thai_H/lesson5/mail_room_part3.py

```diff
@@ -0,0 +1,188 @@
+#!/usr/bin/env python3
+import sys
+import os
+
+
+#--------------------------------------------------------------------------
+def send_thank_you_letter(donor_name, donation_amt):
+    print(f'\nDear {donor_name}\n'
+         'Thank you for a recent donation to our Foundation in the amount of $' + f'{donation_amt}\n'
+         '\nSincerely'
+         '\nBill & Melinda Gates')
+
+
+#---------------------------------------------------------------------------
+def is_existing_donor(donor_name):
+    for each_donor in list_of_donors:
+        if each_donor['name'] == donor_name:
+            return True  # returns True if existing donor
+    return False #False if new donor
+
+
+#-----------------------------------------------------------------------------
+def send_thank_you_add_new_donation():
+    donor_name =input('\n\nPlease enter name of donor:  ')
+
+    try:
+        donation_amt = float(input('\nEnter the amount > '))
+        divide_by_zero = 1 / donation_amt
+    except ZeroDivisionError:
+        print ('Can\'t have zero donation amount')
+        send_thank_you_add_new_donation()
+    else:
+        print ('Amount validated: ', float(donation_amt))
+
+
+    if is_existing_donor(donor_name):
+
+        #--- instead of using this block ------------------------------------
+        #for each_donor in list_of_donors:
+        #    if donor_name == each_donor['name']:
+        #        each_donor['donation'].append(float(donation_amt))
+        #-----------------------------------------------------------------
+
+        #we will use this python list comprehension
+        [each_donor['donation'].append(float(donation_amt)) for each_donor in list_of_donors if donor_name == each_donor['name']]
+
+
+    # New donor
+    else:
+        new_donor_name_amt = {'name':donor_name,'donation':[float(donation_amt)]}
+        list_of_donors.append(new_donor_name_amt)
+    # send thank you letter to the the donor name user just entered
+    send_thank_you_letter(donor_name, donation_amt)
```

Comment:
Adding a new donor is not carrying through when going back to the main menu and seeing the list of donors or creating a report.  Can you see why?